akka-http2-support/src/main/scala/akka/http/impl/engine/http2/AlpnSwitch.scala

```diff
@@ -0,0 +1,121 @@
+    BidiFlow.fromGraph(
+      new GraphStage[BidiShape[HttpResponse, SslTlsOutbound, SslTlsInbound, HttpRequest]] {
+        val netIn = Inlet[SslTlsInbound]("AlpnSwitch.netIn")
+        val netOut = Outlet[SslTlsOutbound]("AlpnSwitch.netOut")
+
+        val requestOut = Outlet[HttpRequest]("AlpnSwitch.requestOut")
+        val responseIn = Inlet[HttpResponse]("AlpnSwitch.responseIn")
+
+        val shape: BidiShape[HttpResponse, SslTlsOutbound, SslTlsInbound, HttpRequest] =
+          BidiShape(responseIn, netOut, netIn, requestOut)
+
+        def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
+          override def preStart(): Unit = pull(netIn)
+
+          setHandler(netIn, new InHandler {
+            def onPush(): Unit =
+              grab(netIn) match {
+                case first @ SessionBytes(session, bytes) ⇒
+                  val chosen = ALPNRegistry.getChosenProtocolAndRemove(WrappedSslContextSPI.engineOfContext(context))
+                  println(s"got bytes for session $session, chosen was $chosen")
+                  chosen match {
+                    case Some("h2") ⇒
+                      println("Selected Http/2")
+                      install(http2Flow, first)
+                    case _ ⇒
+                      println("Selected Http/1")
+                      install(http1Flow, first)
+                  }
+                case SessionTruncated ⇒
+                  println("Got truncated. Ignoring....")
+              }
+          })
+          setHandler(netOut, new OutHandler {
+            def onPull(): Unit = ()
+          })
+          setHandler(requestOut, new OutHandler {
+            def onPull(): Unit = ()
+          })
+          setHandler(responseIn, new InHandler {
+            def onPush(): Unit = throw new IllegalStateException("Wasn't pulled yet")
+          })
+
+          def install(serverImplementation: HttpServerBidiFlow, firstElement: SslTlsInbound): Unit = {
+            val serverDataIn = new SubSinkInlet[SslTlsOutbound]("ServerImpl.netIn")
+            val serverDataOut = new SubSourceOutlet[SslTlsInbound]("ServerImpl.netOut")
+
+            val serverRequestIn = new SubSinkInlet[HttpRequest]("ServerImpl.serverRequestIn")
+            val serverResponseOut = new SubSourceOutlet[HttpResponse]("ServerImpl.serverResponseOut")
+
+            val networkSide = Flow.fromSinkAndSource(serverDataIn.sink, serverDataOut.source)
+            val userSide = Flow.fromSinkAndSource(serverRequestIn.sink, serverResponseOut.source)
+
+            def connect[T](in: Inlet[T], out: SubSourceOutlet[T], initialElement: Option[T]): Unit = {
+              var needToFlushFirst =
+                initialElement match {
+                  case Some(ele) if out.isAvailable ⇒
+                    out.push(ele)
+                    false
+                  case Some(ele) ⇒ true
+                  case None      ⇒ false
+                }
```

Comment:
It's a tiny one-time cost so emit would be fine. Unfortunately Outlet and SubSourceOutlet don't have a common API, so you cannot use `emit` anyways.